tsuru/resource_tsuru_app_autoscale.go

```diff
@@ -0,0 +1,175 @@
+	cpuRegex = regexp.MustCompile(`^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$`)
+)
+
+func resourceTsuruApplicationAutoscale() *schema.Resource {
+	return &schema.Resource{
+		Description:   "Tsuru Application Autoscale",
+		CreateContext: resourceTsuruApplicationAutoscaleCreate,
+		ReadContext:   resourceTsuruApplicationAutoscaleRead,
+		DeleteContext: resourceTsuruApplicationAutoscaleDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"app": {
+				Type:        schema.TypeString,
+				Description: "Application name",
+				Required:    true,
+				ForceNew:    true,
+			},
+			"process": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of service instance",
+			},
+			"min_units": {
+				Type:        schema.TypeInt,
+				Description: "minimum number of units",
+				Required:    true,
+				ForceNew:    true,
+				DefaultFunc: func() (interface{}, error) {
+					return 1, nil
+				},
+			},
+			"max_units": {
+				Type:        schema.TypeInt,
+				Description: "maximum number of units",
+				Required:    true,
+				ForceNew:    true,
+			},
+			"cpu_average": {
+				Type:        schema.TypeString,
+				Description: "cpu average",
+				Required:    true,
+				ForceNew:    true,
+				ValidateFunc: func(i interface{}, k string) (s []string, es []error) {
+					v, ok := i.(string)
+					if !ok {
+						es = append(es, fmt.Errorf("expected type of %s to be string", k))
+						return
+					}
+
+					if !cpuRegex.MatchString(v) {
+						es = append(es, fmt.Errorf("%s value %s must match regexp ^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$", k, v))
+						return
+					}
+					return
+				},
+			},
+		},
+	}
+}
+
+func resourceTsuruApplicationAutoscaleCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	provider := meta.(*tsuruProvider)
+
+	app := d.Get("app").(string)
+
+	appInfo, _, err := provider.TsuruClient.AppApi.AppGet(ctx, app)
+	if err != nil {
+		return diag.Errorf("Unable to read app %s: %v", app, err)
+	}
+
+	if appInfo.Deploys == 0 {
+		return nil
+	}
+
+	process := d.Get("process").(string)
+	minUnits := d.Get("min_units").(int)
+	maxUnits := d.Get("max_units").(int)
+	if minUnits < 0 {
+		minUnits = 1
+	}
+	if maxUnits < 0 {
+		maxUnits = 1
+	}
+	if minUnits > maxUnits {
+		minUnits = maxUnits
+	}
+	cpu := d.Get("cpu_average").(string)
+
+	autoscale := tsuru_client.AutoScaleSpec{
+		Process:    process,
+		MinUnits:   int32(minUnits),
+		MaxUnits:   int32(maxUnits),
+		AverageCPU: cpu,
+	}
+
+	resp, err := provider.TsuruClient.AppApi.AutoScaleAdd(ctx, app, autoscale)
+	if err != nil {
+		return diag.Errorf("Unable to create autoscale %s %s: %v", app, process, err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return diag.Errorf("Unable to create autoscale: error code %d", resp.StatusCode)
+	}
+
+	d.SetId(fmt.Sprintf("%s-%s", app, process))
```

Comment:
could we use `space` as separator of ID ?, `-`, imagine that case:
```
app: myapp-prod, process: web
app: myapp, process: prod-web
```
Will create an unexpected behavior.

Reply:
I think we need a better way of generating unique ids. Not sure what we should use.